### PR TITLE
Make run storage configurable to inmem versus file-based

### PR DESCRIPTION
**Pull request by schrockn**
Introduces the notion of a config-driven decision whether or not to use the filesystem or inmemory for storing runs. The CLI defaults to filesystem; the API called from python defaults to inmemory (for testing). 

This setups the ability to also have an S3-backed intermediates manager and run storage concept.

Next steps on my end:
   - Using this mechanism to replace the execute_marshalling codepath (instead the values will be source from the intermediates manager).
   - Instead of hardcoding to pickling, using the serialization mechanism attached to the type system. Will be needed for spark integration.

### CODE REVIEW

File: python_modules/dagster/dagster/core/definitions/environment_configs.py [hunk 1]
```diff
@@ -217,6 +224,24 @@ def define_expectations_config_cls(name):
     )
 
 
+def define_storage_config_cls(name):
+    check.str_param(name, 'name')
+
+    return SystemNamedSelector(
+        name,
+        {
+            'inmem': Field(
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/definitions/environment_configs.py]: not `in_memory`?

File: python_modules/dagster/dagster/core/definitions/environment_configs.py [hunk 2]
```diff
@@ -323,10 +348,20 @@ def construct_environment_config(config_value):
...
+        storage=construct_storage_config(config_value.get('storage')),
         original_config_dict=config_value,
     )
 
 
+def construct_storage_config(config_value):
+    check.opt_dict_param(config_value, 'config_value', key_type=str)
+    if config_value:
+        storage_key, storage_value = single_item(config_value)
+        return StorageConfig(storage_key, storage_value)
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/definitions/environment_configs.py]: What is this key/value meant to do?

File: python_modules/dagster/dagster/core/execution.py [hunk 3]
```diff
@@ -37,13 +37,18 @@
...
-
 from .errors import DagsterInvariantViolationError
 
 from .events import construct_event_logger
 
 from .execution_plan.create import create_execution_plan_core
 
+from .execution_plan.intermediates_manager import (
+    FileSystemIntermediateManager,
+    InMemoryIntermediatesManager,
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution.py]: `Intermediates` or `Intermediate`

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py [hunk 4]
```diff
@@ -85,8 +86,8 @@ def has_value(self, step_output_handle):
 # will likely be to inject a different type of execution step that threads
 # the manager
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: Do we still think this is true?

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py [hunk 5]
```diff
@@ -107,3 +108,59 @@ def set_value(self, step_output_handle, value):
...
+def check_path_comps(path_comps):
+    path_list = check.list_param(list(path_comps), 'path_comps', of_type=str)
+    check.param_invariant(path_list, 'path_list', 'Must have at least one comp')
+    return path_list
+
+
+class LocalTempFileStore:
+    def __init__(self, run_id):
+        check.str_param(run_id, 'run_id')
+        self.root = os.path.join('/tmp', 'dagster', 'runs', run_id, 'files')
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: Need a FIXME here for Windows

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py [hunk 6]
```diff
@@ -107,3 +108,59 @@ def set_value(self, step_output_handle, value):
...
+        self._created = False
+
+    def ensure_root_exists(self):
+        if not self._created:
+            mkdir_p(self.root)
+
+        self._created = True
```
Comment (mgasner) [inline python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: is this deliberate? if the root is deleted by some other process after `self._created` is set, we'll have errors

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py (hunk 5 above)
Comment (alangenfeld) [reply python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: use `tempfile.gettempdir()`?

**Comment by mgasner** [comment]
Mm, this deserves a changelog entry

File: python_modules/dagster/dagster/core/definitions/environment_configs.py (hunk 1 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/definitions/environment_configs.py]: yup

File: python_modules/dagster/dagster/core/definitions/environment_configs.py (hunk 2 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/definitions/environment_configs.py]: these are poorly named. storage_mode and storage_config are more accurate

File: python_modules/dagster/dagster/core/execution.py (hunk 3 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution.py]: my pluralization, capitalization, and spelling skills continue to plague our codebase

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py (hunk 4 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: some species of this is still true but you're right that the comment is now spurious

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py (hunk 5 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: seven to the rescue

File: python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py (hunk 6 above)
Comment (schrockn) [reply python_modules/dagster/dagster/core/execution_plan/intermediates_manager.py]: yes this really doesn't help that much

**Comment by schrockn** [comment]
This was excellent feedback. Thanks for taking such a close look.

**Pull request merged**